[Meta]: Add CONTRIBUTING.md

ad3154 opened the pull request:
Added basic contibution guidelines and instructions for how to debug the examples using VS Code.

Part of #101

RileyKenyon commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,79 @@
+
+Linux:
+
+Install git, a working CXX compiler, cmake, and git
```
git is written here twice, could replace with "Install the following tools:" to avoid maintenance of writing dependencies multiple times?

ad3154 replied:
Changed to list of tools, with some marked as optional

RileyKenyon commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,79 @@
+
+Install your favorite IDE if you want. We suggest Visual Studio Code because it's easy and we will use it in the remainder of this guide.
+
+VS Code can be installed via the snap store, or by downloading it from Microsoft.
```
```suggestion
VS Code can be installed via the snap store, or by downloading it from [Microsoft](https://code.visualstudio.com/Download).
```

ad3154 replied:
Applied your suggestion

RileyKenyon commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,79 @@
+
+If you want to build the examples and debug with them, you'll need to tell that to CMake.
+
+Go into the CMake Tools extension settings, and add a new configure-time argument for `-DBUILD_EXAMPLES=true`. You may need to scroll down a ways to get to it.
```
I think the other documentation is `-DBUILD_EXAMPLES=ON` may be good to match

ad3154 replied:
Changed to match

RileyKenyon reviewed (changes requested):
Couple comments - thanks for adding it in!

GwnDaan commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,84 @@
+Our style rules and PR reviews are based loosely on Autosar and MISRA suggestions and try to ensure the highest quality possible.
+
+* Contributions must follow the style defined in our `.clang-format` file. You can ensure you pass this check by running `find . -iname *.hpp -o -iname *.cpp | xargs clang-format -i` at the root of the repo.
+* Contributions should follow these additional style requirements, which will be checked in code reviews.
+	* Function names `snake_case`
+	* Variables `camelCase`
+	* Constant values `CAPITALIZED_SNAKE`
+	* Constants on the left in `==` and `!=` checks. Like this: `if (5 == value)` NOT `if (value == 5)`. This is to prevent accidentally omitting an `=` in the operator and creating a runtime bug.
```
Could it be useful to add a line about usage of `std::` here?

jqquanbeck-raven replied:
Furthermore, this should really be a given for any C++ developer, but a note that `using namespace x` should never be in a header file may be good to have

GwnDaan commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,84 @@
+
+You can easily set up a linux PC or virtual machine to develop for this project!
+
+We suggest using Ubuntu, Debian, RHEL, or Raspian, but other systems may also work. Windows development is supported with PEAK CAN hardware devices.
```
Raspbian* instead of Raspian

GwnDaan reviewed (commented):
Had two minor things, other than that it looks good! Thanks

jqquanbeck-raven commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,84 @@
+* Builds must pass the compilation github action
+* `NULL` should not be used when `nullptr` can be used
+* Prefer C++ over C, though exceptions may be granted if needed
+* C++11 or earlier is required to help provide the maximum compatibility to all modern compilers, especially for embedded ones that do not support the most up-to-date versions of the standard, such as ARM Compiler 5 and 6.
```
ARM Compiler V6 fully supports C++14. May be somewhat freeing to drop support for V5 which doesn't even fully support C++11.

ad3154 replied:
Yeah, I can get on-board with that. 5 is technically EOL I think so I am ok with raising it to 14

ad3154 commented:
Added and clarified some style items, such as:

* Explicit namespacing should be used for accessing namespaces outside of our namespace `isobus`, especially for the standard library `std::`
* No `using namespace` directives are allowed in header files
* Constructors with a single parameter shall be marked `explicit`
* Where possible, append `const` to functions that do not modify class member variables
* Where possible, declare variables as `constexpr` if their value can be determined at compile time
* Only those characters specified in the C++ Language Standard basic source character set shall be used in the source code except within the text of a wide string.

Fixed spelling of Raspbian
Updated CXX standard to 14 by default. I had originally set it to 11 to try and support ARMCC (ARM Compiler 5) but to be honest, that compiler is EOL and doesn't support the full C++11 standard anyways so it may not have worked to begin with.

RileyKenyon reviewed (approved):
Looks good to me!
